### PR TITLE
BE- Add Event Type endpoints:

### PR DESCRIPTION
**Pull request by viriponce**
This PR addresses issue #23 

**Implementation**

1. **backend/api/models.py**
Implemented the EventType model with the following fields:

- stroke:  = models.CharField(max_length=20, choices=Stroke.choices, blank=False, null=False)
- distance = models.PositiveSmallIntegerField(blank=False, null=False)
- type = models.CharField(max_length=20, choices=Type.choices, blank=False, null=False, default=Type.INDIVIDUAL)

Introduced a name property to dynamically generate the event type name based on the three fields mentioned above.

Create an unique constraint for the combinations of those three fields

2. **backend/api/migrations/0009_eventtype_eventtype_unique_event_type.py**
After creating the model, I ran the `makemigrations` command and the migration for the EventType model was created.

4.  **backend/api/serializers/EventTypeSerializer.py**
Created a Model serializer for the EventType model. Included all the fields plus the name.
Defined stroke and type as ChoiceField in order to validate the values passed for these against the choices defined in the model for them.
```
stroke = serializers.ChoiceField(choices=EventType.Stroke.choices,required=True)
type = serializers.ChoiceField(choices=EventType.Type.choices, required=True)
```

Incorporated the following validations:

- `validate_distance` : Field validator to return a bad request when the value provided for distance is negative or equal to 0.
-` validate_unique_event_type`: to return a bad request status code when the combinatios of the values for distance, stroke and type already exists.
- `validate` method override: call the  validate_unique_event_type validator.

5. **backend/api/views/EventTypeView.py**
Created a basic Model View  set for EventType.

6. **backend/api/urls.py**
Added a router to generate the endpoints for Event Type.

7. **backend/api/fixtures/setup_data.json**

Added data for 9 event type objects.

I applied the migration with `migrate` command, then loaded the fixture data with `loaddata` command. 

**Swagger**

The endpoints are accessible through Swagger:
Screenshot 2024-02-15 at 1 41 49 PM

I tried each of the new endpoints successfully. I was able to create, delete, patch and get an event type.
When posting and patching, I got the expected error for entering invalid data.

### CODE REVIEW

File: backend/api/fixtures/setup_data.json [hunk 1]
```diff
@@ -90,5 +90,87 @@
           "min_age": 11,
           "max_age": null
       }
+  },
+  {
+    "model": "api.eventtype",
+    "pk": 1,
+    "fields": {
+        "distance": 50,
+        "stroke": "M",
```
Comment (MarisiaS) [inline backend/api/fixtures/setup_data.json:99]: Could you change this one to Fly? please.
A 50 Medley is not posible and also we don't have any fly event.

File: backend/api/serializers/EventTypeSerializer.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+            distance=distance,
+            stroke=stroke,
+            type=type
+            ).exists():
+                raise serializers.ValidationError("An Event Type with these caracteristics already exists.")
+        return data
+    
+    def validate_distance(self, value):
+        if value <= 0:
+            raise serializers.ValidationError("Distance must be a positive number greater than zero.")
```
Comment (MarisiaS) [inline backend/api/serializers/EventTypeSerializer.py:35]: I think "positive" is not need because  'greater than zero' already implies positivity.

**Review by MarisiaS — COMMENTED** [review]
We've observed that the requirement states 'Distance must be greater than 0,' yet the examples in both the request and response show the distance being set to 0. It might be beneficial to update the examples accordingly.
image

image

**Comment by viriponce** [comment]
> We've observed that the requirement states 'Distance must be greater than 0,' yet the examples in both the request and response show the distance being set to 0. It might be beneficial to update the examples accordingly. image
> 
> image

For me, the distance had a different default value:
Screenshot 2024-02-16 at 9 07 55 AM

Screenshot 2024-02-16 at 9 07 23 AM

I'm trying to  explicitly created the example for those methods but I'm not getting the expected result yet.

**Review by MarisiaS — APPROVED** [review]
LGTM

The evnt type table meets the requirements.
image

I was able to load the fixture data
image

I tried the endpoints in Swagger, I was able to get an specific event type by id, delete it, patch it.
For the post endpoint, I tried different combinations for the parameters and got the expected error messages when providing invalid data.
A new record is created when providing valid data.

Thanks Viri!

**Pull request merged**